Store uncompressed AIPs

mistydemeo opened the pull request:
This is the sister PR to artefactual/archivematica#40.

This allows storing uncompressed AIPs via the storage service.
- Store AIP now works if no pointer file is present.
- Packages can now compress themselves, and the download API from the package resource will compress uncompressed packages before serving them.

Ignore the last commit here; it's duplicated from the other branch.

Commit:
Store AIP: handle AIPs without pointer files

This is needed to support uncompressed AIPs, which do not have
pointer files.

refs #6589

Commit:
Package: add new compress_package instance method

This supports features such as downloading entire AIPs for
uncompressed AIPs.

Note that this currently only tarballs the file without performing
actual compression.

refs #6589

mistydemeo committed:
Package: don't try to call methods on None pointer_file_location

This will be unset for certain types of packages, like uncompressed AIPs.

refs #6589

Hwesta commented on storage_service/locations/models.py:
```diff
@@ -1281,8 +1285,14 @@ def store_aip(self, origin_location, origin_path):
+            src_space.move_to_storage_service(pointer_file_src, pointer_file_name, self.pointer_file_location.space)
+            self.pointer_file_location.space.move_from_storage_service(pointer_file_name, pointer_file_dst)
+        except:
+            logging.warn("No pointer file found")
```
`logging.warn` has been [deprecated by removing docs about it](http://bugs.python.org/issue13235), and replaced by `logging.warning`

mistydemeo replied:
Thanks, fixed.

Hwesta commented on storage_service/locations/models.py:
```diff
@@ -1203,6 +1203,10 @@ def is_compressed(self):
         full_path = self.full_path()
         if lockss_au_number is None:
+            if not self.is_compressed():
+                path, _ = self.compress_package()
```
The temp dir this is compressed to should be deleted.  If the temp dir path can make it to `utils.download_file_stream`'s temp_dir parameter, that will delete it.

mistydemeo replied:
Decided I didn't like pushing this into `get_download_path` anyway (I definitely wouldn't expect `get_download_path` to churn away on 500 gigs of data!), so made it raise an exception in that case and had the resource handle compressing the file and passing the temp dir as appropriate.

Hwesta commented on storage_service/locations/models.py:
```diff
@@ -1180,7 +1180,7 @@ def full_pointer_file_path(self):
 
         Includes the space, location and package paths joined."""
-        if self.package_type not in (self.AIP, self.AIC):
+        if not self.pointer_file_location:
```
`resources.PackageResource.pointer_file_request` should be updated to handle the path being None, since it's also called from AM (archival storage tab).

mistydemeo replied:
Pushed a commit to cause it to return 404 if no pointer file exists.

mistydemeo committed:
Pointer file request: return 404 if no pointer file exists

refs #6589

Hwesta commented on storage_service/locations/models.py:
```diff
@@ -1376,6 +1385,39 @@ def extract_file(self, relative_path='', extract_path=None):
+            # The filenames are transformed to ensure only the
+            # directory basename is present, not the entire path.
+            # FIXME This syntax isn't portable and is only compatible with GNU tar
+            "--xform", "s,{}/,,".format(os.path.dirname(full_path[1:-1])),
```
As mentioned in person, this can be simplified to `command = ['tar', 'c', '-C', relative_path, '-f', compressed_filename, basename]` where `relative_path = ''.join(full_path.rsplit(basename, 1))` (replace once, but starting from the right).  You can update here, or I can update in the AIP reingest branch.

Hwesta commented on storage_service/locations/models.py:
```diff
@@ -1203,6 +1203,8 @@ def is_compressed(self):
     def get_download_path(self, lockss_au_number=None):
         full_path = self.full_path()
         if lockss_au_number is None:
+            if self.is_compressed():
```
Should be `if not self.is_compressed():`

Hwesta commented:
`PackageResource.extract_file_request` needs to be updated as well (though I think the problem is actually that extract_file doesn't append the relative_path instead of the basename if the AIP is uncompressed).

Commit:
Extract file request: adjust path for uncompressed packages

refs #6589

Commit:
Extract file request: return 404 if requested file not present

refs #6589

Commit:
Extract file request: return 501 if package can't be extracted

refs #6589

mistydemeo commented:
Updated branch with a fix to extract_file_request to adjust the path, and a couple of other commits to return correct status codes and useful error messages in failure cases.

Hwesta commented:
:+1:

mistydemeo committed:
PackageResource: allow downloading uncompressed packages

This also causes Package.get_download_path() to raise, as uncompressed
directories are not considered downloadable.

refs #6589
